File: solutions/code/206. Reverse Linked List/solution2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode reverseListHelper(ListNode cur, ListNode prev) {
        if (cur == null) {
            return prev;
        }

        ListNode next = cur.next;
        cur.next = prev;
        return reverseListHelper(next, cur);
    }

    public ListNode reverseList(ListNode head) {
        return reverseListHelper(head, null);
    }
}

File: code/0206. Reverse Linked List/solution2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode* reverseListHelper(ListNode* cur, ListNode* prev) {
        if (!cur) {
            return prev;
        }

        ListNode* next = cur->next;
        cur->next = prev;
        return reverseListHelper(next, cur);
    }

public:
    ListNode* reverseList(ListNode* head) {
        return reverseListHelper(head, nullptr);
    }
};